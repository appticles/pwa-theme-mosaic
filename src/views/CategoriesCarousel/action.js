import config from 'react-global-configuration';

export const REQUEST_CATEGORIES = 'REQUEST_CATEGORIES';
export const RECEIVE_CATEGORIES = 'RECEIVE_CATEGORIES';

export const requestCategories = () => ({
  type: REQUEST_CATEGORIES,
});

export const receiveCategories = categories => ({
  type: RECEIVE_CATEGORIES,
  categories,
});

export const fetchCategories = (params = {}) => (dispatch) => {
  dispatch(requestCategories());

  let url;
  if (params && params.post) {
    url = config.get('export').categories + '?post=' + String(params.post);
  } else {
    url =
      config.get('export').categories +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receiveCategories(json)))
    .catch(() => {
      dispatch(receiveCategories([]));
    });
};
