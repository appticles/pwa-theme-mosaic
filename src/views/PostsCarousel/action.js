import config from 'react-global-configuration';

export const REQUEST_POSTS = 'REQUEST_POSTS';
export const RECEIVE_POSTS = 'RECEIVE_POSTS';

export const requestPosts = () => ({
  type: REQUEST_POSTS,
});

export const receivePosts = posts => ({
  type: RECEIVE_POSTS,
  posts,
});

export const fetchPosts = (params = {}) => (dispatch) => {
  dispatch(requestPosts());
  config.set(window.__INITIAL_CONFIG__, { freeze: false });

  let url;
  if (params && params.id) {
    url = config.get('export').posts + '/' + String(params.id);
  } else {
    url =
      config.get('export').posts +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receivePosts(json)))
    .catch(() => {
      dispatch(receivePosts([]));
    });
};
