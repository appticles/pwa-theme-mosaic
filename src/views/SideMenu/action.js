import config from 'react-global-configuration';

export const REQUEST_PAGES = 'REQUEST_PAGES';
export const RECEIVE_PAGES = 'RECEIVE_PAGES';

export const requestPages = () => ({
  type: REQUEST_PAGES,
});

export const receivePages = pages => ({
  type: RECEIVE_PAGES,
  pages,
});

export const fetchPages = (params = {}) => (dispatch) => {
  dispatch(requestPages());

  let url;
  if (params && params.id) {
    url = config.get('export').pages + '/' + String(params.id);
  } else {
    url =
      config.get('export').pages +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receivePages(json)))
    .catch(() => {
      dispatch(receivePages([]));
    });
};
