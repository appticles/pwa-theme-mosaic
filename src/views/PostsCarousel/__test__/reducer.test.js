import posts, { INITIAL_STATE, getPosts, getPostsFetching, getLoadMorePosts, getPostsByCategory, getPostProtectedStatus } from '../reducer';
import { requestPosts, receivePosts } from '../action';

describe('Posts reducer', () => {
  it('returns same state on an unhandled action', () => {
    expect(posts(INITIAL_STATE, { type: '_NULL' })).toMatchSnapshot();
  });

  it('handles REQUEST_POSTS action', () => {
    const postsList = {
      isFetching: 0,
      items: [
        {
          id: 1,
          name: 'Nice post',
        },
        {
          id: 2,
          name: 'Nicer post',
        },
      ],
      loadMore: false,
    };
    expect(posts(postsList, requestPosts())).toMatchSnapshot();
  });

  it('handles RECEIVE_POSTS action and adds new posts to the state', () => {
    const postState = {
      isFetching: 1,
      items: [
        {
          id: 1,
          name: 'Nice post',
        },
        {
          id: 2,
          name: 'Nicer post',
        },
      ],
      loadMore: true,
    };

    const response = [
      {
        id: 3,
        name: 'Nicest post',
      },
    ];

    expect(posts(postState, receivePosts(response))).toMatchSnapshot();
  });

  it('handles RECEIVE_POSTS action and merges by post id', () => {
    const postState = {
      isFetching: 1,
      items: [
        {
          id: 1,
          name: 'Nice post',
        },
        {
          id: 2,
          name: 'Nicer post',
        },
      ],
      loadMore: true,
    };

    const response = [
      {
        id: 2,
        name: 'Nicer post',
      },
    ];

    expect(posts(postState, receivePosts(response))).toMatchSnapshot();
  });

  it('handles RECEIVE_POSTS action and continues load more', () => {
    const postState = {
      isFetching: 1,
      items: [
        { id: 1 },
        { id: 2 },
      ],
      loadMore: true,
    };

    const response = [
      { id: 1 },
      { id: 2 },
      { id: 3 },
      { id: 4 },
      { id: 5 },
      { id: 6 },
      { id: 7 },
      { id: 8 },
      { id: 9 },
      { id: 10 },
    ];

    expect(posts(postState, receivePosts(response))).toMatchSnapshot();
  });

  // reducer functions
  it('getPostsFetching returns state.isFetching', () => {
    expect(getPostsFetching(INITIAL_STATE)).toBe(0);
  });

  it('getLoadMorePosts returns state.loadMore', () => {
    const state = {
      loadMore: true,
    };
    expect(getLoadMorePosts(state)).toBe(true);
  });

  it('getPosts returns state.items', () => {
    const state = {
      items: [
        {
          name: 'Some post',
        },
      ],
    };
    expect(getPosts(state)).toBe(state.items);
  });

  it('getPostsByCategory returns posts from category', () => {
    const state = {
      items: [
        {
          name: 'Some post',
          categories: [101, 102],
        },
        {
          name: 'Another post',
          categories: [102, 103],
        },
      ],
    };
    expect(getPostsByCategory(state, 101)).toMatchSnapshot();
    expect(getPostsByCategory(state, 102)).toMatchSnapshot();
    expect(getPostsByCategory(state, 104)).toMatchSnapshot();
  });

  it('getPostProtectedStatus returns true for password protected post and false otherwise', () => {
    const state = {
      items: [
        {
          id: 1,
          name: 'Password protected post',
          content: {
            protected: true,
          },
        },
        {
          id: 2,
          name: 'Public post',
          content: {
            protected: false,
          },
        },
      ],
    };
    expect(getPostProtectedStatus(state, 1)).toBe(true);
    expect(getPostProtectedStatus(state, 2)).toBe(false);

    // non existing post returns a protected status
    expect(getPostProtectedStatus(state, 3)).toBe(true);
    expect(getPostProtectedStatus(INITIAL_STATE, 3)).toBe(true);
  });
});
