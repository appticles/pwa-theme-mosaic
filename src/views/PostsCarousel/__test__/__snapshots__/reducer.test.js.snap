// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts reducer getPostsByCategory returns posts from category 1`] = `
Array [
  Object {
    "categories": Array [
      101,
      102,
    ],
    "name": "Some post",
  },
]
`;

exports[`Posts reducer getPostsByCategory returns posts from category 2`] = `
Array [
  Object {
    "categories": Array [
      101,
      102,
    ],
    "name": "Some post",
  },
  Object {
    "categories": Array [
      102,
      103,
    ],
    "name": "Another post",
  },
]
`;

exports[`Posts reducer getPostsByCategory returns posts from category 3`] = `Array []`;

exports[`Posts reducer handles RECEIVE_POSTS action and adds new posts to the state 1`] = `
Object {
  "isFetching": 0,
  "items": Array [
    Object {
      "id": 1,
      "name": "Nice post",
    },
    Object {
      "id": 2,
      "name": "Nicer post",
    },
    Object {
      "id": 3,
      "name": "Nicest post",
    },
  ],
  "loadMore": false,
}
`;

exports[`Posts reducer handles RECEIVE_POSTS action and continues load more 1`] = `
Object {
  "isFetching": 0,
  "items": Array [
    Object {
      "id": 1,
    },
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
    Object {
      "id": 5,
    },
    Object {
      "id": 6,
    },
    Object {
      "id": 7,
    },
    Object {
      "id": 8,
    },
    Object {
      "id": 9,
    },
    Object {
      "id": 10,
    },
  ],
  "loadMore": true,
}
`;

exports[`Posts reducer handles RECEIVE_POSTS action and merges by post id 1`] = `
Object {
  "isFetching": 0,
  "items": Array [
    Object {
      "id": 1,
      "name": "Nice post",
    },
    Object {
      "id": 2,
      "name": "Nicer post",
    },
  ],
  "loadMore": false,
}
`;

exports[`Posts reducer handles REQUEST_POSTS action 1`] = `
Object {
  "isFetching": 1,
  "items": Array [
    Object {
      "id": 1,
      "name": "Nice post",
    },
    Object {
      "id": 2,
      "name": "Nicer post",
    },
  ],
  "loadMore": true,
}
`;

exports[`Posts reducer returns same state on an unhandled action 1`] = `
Object {
  "isFetching": 0,
  "items": Array [],
  "loadMore": true,
}
`;
